prow/github/client_test.go

```diff
@@ -3237,3 +3278,42 @@ func TestCreatePullRequestReviewComment(t *testing.T) {
 		t.Errorf("Didn't expect error: %v", err)
 	}
 }
+
+func TestThrottlerRespectsContexts(t *testing.T) {
+	ts := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(200)
+	}))
+	defer ts.Close()
+	c := getClient(ts.URL)
+
+	// Set the throttler, use up the one token we have for this hour
+	c.Throttle(1, 1)
+	if err := c.CreateReview("", "", 0, DraftReview{}); err != nil {
+		t.Fatalf("failed to use up the throttlers token: %v", err)
+	}
+
+	// Use a very low timeout so we don't have to wait
+	ctx, cancel := context.WithTimeout(context.Background(), time.Millisecond)
+	defer cancel()
+
+	clientMethods := getCallForAllClientMethodsThroughReflection(c,
+		// Skip all methods whose name doesn't end with 'WithContext'
+		func(methodName string) bool { return !strings.HasSuffix(methodName, "WithContext") },
```

Comment:
While this is clever I do wonder if there's a more robust way to use the type system to confer this information rather than checking method name suffixes for strings and "type checking" with strings as well.

Reply:
Hm. Maybe test if the method is exported and takes a context.Context as first arg?

Reply:
Or perhaps a second interface that has only context-using methods? I'm not really sure what's best

Reply:
> Or perhaps a second interface that has only context-using methods? I'm not really sure what's best

That is difficult, because the github client is already divided in a bunch of sub interfaces. Additionally, relying on someone adding this to the right interface is no less error-prone than relying on them adhering to a namescheme.

I've changed this to just test all methods whose first argument is of type `context.Context`. IMHO this is as good as this will ever get.